add code and tests to detect vboxnet collision with host interfaces

paul-callahan opened the pull request:
With the virtualbox machine driver, if one of the host's network interfaces has the same subnet/broadcast as the host-only virtualbox network, there will be problems creating and accessing the docker-machine vm(s).  The only indication of the problem is a "no route to host" error in the logs.  Virtualbox does not complain if a host-only network is created that conflicts with a host interface.  

This change will raise an error if docker-machine attempts to create or start a vm that has a host-only network with a CIDR that overlaps/collides with an interface on the host.  

unit tests pass and core-commands.bats passes with virtualbox driver.

 Fixes #3060 

Signed-off-by: Paul Callahan <paul.callahan@gmail.com>

nathanleclaire commented:
Nice!  Thanks for the PR.   Will take a look.

dgageot commented:
LGTM

nathanleclaire commented:
Hm, I get an error trying a `create` with the PR, are these host only interfaces are invalid?

```
$ VBoxManage list hostonlyifs
Name:            vboxnet0
GUID:            786f6276-656e-4074-8000-0a0027000000
DHCP:            Disabled
IPAddress:       192.168.99.1
NetworkMask:     255.255.255.0
IPV6Address:
IPV6NetworkMaskPrefixLength: 0
HardwareAddress: 0a:00:27:00:00:00
MediumType:      Ethernet
Status:          Up
VBoxNetworkName: HostInterfaceNetworking-vboxnet0

Name:            vboxnet1
GUID:            786f6276-656e-4174-8000-0a0027000001
DHCP:            Disabled
IPAddress:       192.168.59.3
NetworkMask:     255.255.255.0
IPV6Address:
IPV6NetworkMaskPrefixLength: 0
HardwareAddress: 0a:00:27:00:00:01
MediumType:      Ethernet
Status:          Down
VBoxNetworkName: HostInterfaceNetworking-vboxnet1
```

nathanleclaire commented on drivers/virtualbox/network.go:
```diff
@@ -333,6 +333,44 @@ func listDHCPServers(vbox VBoxManager) (map[string]*dhcpServer, error) {
+
+// checkIPNetCollison returns true if any host interfaces conflict with the host-only network mask passed as a parameter.
+// This works with IPv4 or IPv6 ip addresses.
+func checkIPNetCollison(hostonly *net.IPNet, hostIfaces map[string]*net.IPNet) (bool, error) {
```
Nitpick: It should be 'checkIPNetCollision` (missing an `i`)

scmadmin commented:
@nathanleclaire I wasn't excluding existing host-only nets, so that error occured.   I fixed that and added an extra test case.

To exclude the existing host-onlies, I had to call listHostOnlyAdapters... so to prevent this from being called twice in a row (and also messing up the order of the vbox mock calls), I had to take that method out of getOrCreateHostOnlyNetwork (and I updated the tests to reflect this).   I'll make some notes in the review.

Hope this is ok.

Also, one other thing - I'm not mocking the calls to net.Interfaces(), etc, so the tests might pick up some interfaces if the host running the tests have some existing virtualbox host only networks - this could cause spurious test errors.   Is this ok, or should I try to mock the calls to net.* ?

paul-callahan commented:
@nathanleclaire I wasn't excluding existing host-only nets, so that error occured.   I fixed that and added an extra test case.

To exclude the existing host-onlies, I had to call listHostOnlyAdapters... so to prevent this from being called twice in a row (and also messing up the order of the vbox mock calls), I had to take that method out of getOrCreateHostOnlyNetwork (and I updated the tests to reflect this).   I'll make some notes in the review.

Hope this is ok.

Also, one other thing - I'm not mocking the calls to net.Interfaces(), etc, so the tests might pick up some interfaces if the host running the tests have some existing virtualbox host only networks - this could cause spurious test errors.   Is this ok, or should I try to mock the calls to net.* ?

scmadmin commented on drivers/virtualbox/network.go:
```diff
@@ -158,12 +158,7 @@ func getHostOnlyAdapter(nets map[string]*hostOnlyNetwork, hostIP net.IP, netmask
-	if err != nil {
-		return nil, err
-	}
-
```
moved listHostOnlyAdapters() up one level to prevent VBoxManager from being called twice in a row with the same command.

paul-callahan replied:
moved listHostOnlyAdapters() up one level so this wouldn't be called twice in a row (resulting in two exact same calls to VBoxManage in a row)

nathanleclaire commented:
> Also, one other thing - I'm not mocking the calls to net.Interfaces(), etc, so the tests might pick up some interfaces if the host running the tests have some existing virtualbox host only networks - this could cause spurious test errors. Is this ok, or should I try to mock the calls to net.* ?

Please mock them.  Thanks!

nathanleclaire commented:
Great.  @paul-callahan is this ready for final review / merge?

paul-callahan commented:
Yup, ready.  Thanks.

On Fri, Mar 11, 2016 at 12:20 PM, Nathan LeClaire <notifications@github.com>
wrote:

> Great. @paul-callahan <https://github.com/paul-callahan> is this ready
> for final review / merge?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/machine/pull/3165#issuecomment-195530715>.
>

nathanleclaire commented:
Create works for me now :+1: 

LGTM

Nice work on this!
